qiskit/circuit/qpy_serialization.py

```diff
@@ -0,0 +1,706 @@
+the circuit.
+
+REGISTERS
+---------
+
+The contents of REGISTERS is a number of REGISTER object. If num_registers is
+> 0 then after reading METADATA you read that number of REGISTER structs defined
+as:
+
+.. code-block:: c
+
+    struct {
+        char type;
+        unsigned int size;
+        char name[10];
+    }
+
+``type`` can be ``'q'`` or ``'c'``.
+
+CUSTOM_DEFINITIONS
+------------------
+
+If the circuit contains custom defitions for any of the instruction in the circuit.
+this section
+
+CUSTOM_DEFINITION_HEADER contents are defined as:
+
+.. code-block:: c
+
+    struct {
+        unsigned long long size;
+    }
+
+If size is greater than 0 that means the circuit contains a custom instruction.
+Each custom instruction is defined with a CUSTOM_INSTRUCTION block defined as:
+
+.. code-block:: c
+
+    struct {
+        char name[32];
+        char type;
+        _Bool custom_definition;
+        unsigned long long size
+    }
+
+If ``custom_definition`` is ``True`` that means that the immediately following
+` size`` bytes contains a QPY circuit data which can be used for the custom
+definition of that gate name. If ``custom_definition`` is ``False`` than the
+instruction can be considered opaque (ie no definition).
+
+INSTRUCTIONS
+------------
+
+The contents of INSTRUCTIONS is a list of INSTRUCTION metadata objects
+
+.. code-block:: c
+
+    struct {
+        char name[32];
+        unsigned short num_parameters;
+        unsigned int num_qargs;
+        unsigned int num_cargs;
+        _Bool has_conditionl
+        char conditonal_reg[10];
+        long long conditional_value;
+    }
+
+``name`` here is the Qiskit class name for the Instruction class if it's
+defined in Qiskit. Otherwise it falls back to the custom instruction name.
+
+which is immediately followed by the INSTRUCTION_ARG structs for the list of
+arguments of that instruction. These are in the order of all quantum arguments
+(there are num_qargs of these) followed by all classical arguments (num_cargs
+of these).
+
+The contents of each INSTRUCTION_ARG is:
+
+.. code-block:: c
+
+    struct {
+        char type;
+        unisgned int size;
+        char name[10];
+    }
+
+``type`` can be ``'q'`` or ``'c'``.
+
+After all arguments for an instruction the parameters are specified with
+``num_parameters`` INSTRUCTION_PARAM structs.
+
+The contents of each INSTRUCTION_PARAM is:
+
+.. code-block:: c
+
+    struct {
+        char type;
+        unsigned long long size;
+    }
+
+After each INSTRUCTION_PARAM the next ``size`` bytes are the parameter's data.
+The ``type`` field can be ``'i'``, ``'f'``, ``'p'``, or ``'n'`` which dictate
+the format. For ``'i'`` it's an integer, ``'f'`` it's a double, ``'p'`` defines
+a :class:`~qiskit.circuit.Paramter` object  which is represented by a PARAM
+struct (see below), and ``'n'`` represents an object from numpy (either an
+``ndarray`` or a numpy type) which means the data is .npy format [#f2]_
+data.
+
+
+PARAMETER
```

Comment:
Parameter expressions aren't supported yet (it's listed somewhere, maybe the PR summary, as a current limitation). The reason is since the parameter expressions are fundamentally just a sympy expression and I would basically need a way to serialize the sympy expression tree which seemed doable but not trivial for this first iteration.